config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/util/environment/EnvironmentConfigurationUtils.java

```diff
@@ -36,40 +42,83 @@
+        entityFileMap.put(EntityTypes.LISTEN_PORT_TYPE, ImmutablePair.of("LISTEN_PORT","listen-ports.yml"));
+        entityFileMap.put(EntityTypes.STORED_PASSWORD_TYPE, ImmutablePair.of("PASSWORD","stored-passwords.properties"));
+        entityFileMap.put(EntityTypes.CLUSTER_PROPERTY_TYPE, ImmutablePair.of("PROPERTY","global-env.properties"));
+
+        return entityFileMap;
     }
 
-    private String getEnvValue(String key, Object o) {
-        if (o instanceof String) {
-            return (String) o;
+    /**
+     * Parses the deployment bundle metadata file to extract environmental dependencies.
+     *
+     * @param metaDataFile deployment bundle metadata.
+     * @param configFolder config folder to look for environmental files.
+     * @return pair of bundle file name and environmental entities map.
+     */
+    public Pair<String, Map<String, String>> parseBundleMetadata(File metaDataFile, File configFolder) {
+        if (!metaDataFile.exists()) {
+            throw new MissingEnvironmentException("Metadata file " + metaDataFile.toString() + " does not exist.");
         }
-        if (o instanceof File) {
-            // get the entity type and name
-            String entityType = key.substring(0, key.indexOf('.'));
-            String entityName = key.substring(key.indexOf('.') + 1);
-            return loadConfigFromFile((File) o, entityType, entityName);
+        final EnvironmentBundleData environmentBundleData = jsonFileUtils.readBundleMetadataFile(metaDataFile, EnvironmentBundleData.class);
+        if (environmentBundleData != null && environmentBundleData.getEnvironmentEntities() != null && !environmentBundleData.getEnvironmentEntities().isEmpty()) {
+            final String bundleName = environmentBundleData.getName();
+            final String bundleVersion = environmentBundleData.getVersion();
+            final Map<String, String> environmentValues = new LinkedHashMap<>();
+            final List<Map<String, String>> environmentEntities = environmentBundleData.getEnvironmentEntities();
+            environmentEntities.stream().forEach(environmentEntitiy -> {
+                final String entityType = environmentEntitiy.get("type");
+                final String entityName = !EntityTypes.CLUSTER_PROPERTY_TYPE.equals(entityType) ? environmentEntitiy.get("name") : PREFIX_GATEWAY + environmentEntitiy.get("name");
+                final Pair<String, String> entityFilePair = ENTITY_FILE_MAP.get(entityType);
+                if (null == entityFilePair) {
+                    throw new MissingEnvironmentException("Unexpected entity type " + entityType);
+                }
+
+                final File envConfigFile = new File(configFolder, entityFilePair.getRight());
+                environmentValues.put(PREFIX_ENV + entityFilePair.getLeft() + "." + entityName,
+                        loadConfigFromFile(envConfigFile, entityFilePair.getLeft(), entityName));
```

Comment:
does policy refer any environment entity by id or guid ? if yes there will be mismatch between deployment bundle and environment bundle

Reply:
both are different bundles, environment entities will go first and then we are mapping my name, i do not think there will be an issue

Reply:
In Trusted_cert case there is a mismatch.
During the export, DEV PLUGIN is ignoring the entity ID and building the entity with a new ID as part of Env bundle, but Policy entity still builds with the old XML which has trusted cert ID as reference. Created defect [DE461108](https://rally1.rallydev.com/#/40967354777d/detail/defect/391587169100?fdp=true)